src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs

```diff
@@ -75,42 +92,68 @@ private async Task CreateBlocksAsync()
             {
                 try
                 {
-                    // Don't mine in IBD.
-                    if (this.ibdState.IsInitialBlockDownload() && !MineDuringIBD)
+                    // Don't mine in IBD in case we are connected to any node.
+                    if (this.ibdState.IsInitialBlockDownload() && this.connectionManager.ConnectedPeers.Any())
                     {
                         int attemptDelayMs = 20_000;
                         await Task.Delay(attemptDelayMs, this.cancellation.Token).ConfigureAwait(false);
 
                         continue;
                     }
 
+                    uint timeNow = (uint) this.dateTimeProvider.GetAdjustedTimeAsUnixTimestamp();
+                    uint myTimestamp = this.slotsManager.GetMiningTimestamp(timeNow);
+
+                    if (myTimestamp < timeNow)
+                    {
+                        // Sanity check.
+                        throw new Exception("Invalid mining timestamp generated.");
+                    }
 
-                    // TODO POA check if our timestamp and if not wait for our timestamp (wait with cancellation token)
+                    uint waitingTime = myTimestamp - timeNow - 1;
 
-                    // TODO will set nonce and target to const. Check todo in PoABlockDefenition
+                    if (waitingTime > 0)
+                    {
+                        // Wait till we can mine.
+                        this.logger.LogInformation("Waiting {0} seconds before block can be mined.", waitingTime);
```

Comment:
We can even have this outside the if. "Waiting 0 seconds before .. " is also a useful message.
Also, before block -> until a block.

Reply:
Fixed